controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,21 +140,215 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
+				}
+				// create a new cloudstack machine template, which will be referred by etcdadmcluster spec.infrastructureTemplate
+				csMachineTemplate.Name = fmt.Sprintf("%s-template-%d", ref.Name, time.Now().UnixNano()/int64(time.Millisecond))
+				csMachineTemplate.SetAnnotations(map[string]string{})
+				csMachineTemplate.SetResourceVersion("")
+				if err := r.K8sClient.Create(r.RequestCtx, csMachineTemplate); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// patch etcdadmcluster to refer newly created cloudstack machine template, which will trigger an etcdadmcluster rolling upgrade.
+				if _, err := controllerutil.CreateOrPatch(r.RequestCtx, r.K8sClient, etcdadmCluster,
+					func() error {
+						return unstructured.SetNestedField(
+							etcdadmCluster.Object,
+							csMachineTemplate.Name,
+							"spec", "infrastructureTemplate", "name",
+						)
+					}); err != nil {
+					return ctrl.Result{}, err
+				}
+				r.Log.Info(fmt.Sprintf("failuredomain delete: etcdadmcluster infrastructureTemplate cloudstackmachinetemplate name %s patched", csMachineTemplate.Name))
+				break
 			}
 		}
+		if etcdMachineFound {
+			break
+		}
 	}
-	if len(machines.Items) > 0 {
-		return r.RequeueWithMessage("FailureDomain still has machine(s) in it.")
+
+	return ctrl.Result{}, nil
+}
+
+func triggerMachineDeploymentRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	workerMachineFound := false
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == "MachineSet" {
+				workerMachineFound = true
+
+				// get machine deployment by using cloudstack machine's label info
+				md := &clusterv1.MachineDeployment{}
+				mdName, ok := machine.Labels[clusterv1.MachineDeploymentLabelName]
+				if !ok {
+					return ctrl.Result{}, errors.Errorf("cloudstack machine %s misses label: %s", machine.Name, clusterv1.MachineDeploymentLabelName)
+				}
+
+				if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: machine.Namespace, Name: mdName}, md); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// add an annotation restartedAt in machine deployment if such one not already added
+				// this will trigger an immediate machine deployment rollout
+				_, ok = md.Spec.Template.Annotations["cluster.x-k8s.io/restartedAt"]
+				if !ok {
+					if md.Spec.Template.Annotations == nil {
+						md.Spec.Template.Annotations = map[string]string{}
+					}
+					timeNowStr := time.Now().Format(time.RFC3339)
+					md.Spec.Template.Annotations["cluster.x-k8s.io/restartedAt"] = timeNowStr
+					patcher, err := patch.NewHelper(md, r.K8sClient)
+					if err != nil {
+						return ctrl.Result{}, err
+					}
+					if err := patcher.Patch(r.RequestCtx, md); err != nil {
+						return ctrl.Result{}, err
+					}
+					r.Log.Info(fmt.Sprintf("failuredomain delete: machine deployment name %s add annotation restartedAt %s", mdName, timeNowStr))
```

Comment:
how does this work when a cluster has multiple worker node groups? Has that been tested at all?